compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -506,6 +508,33 @@ fn layout_of_uncached<'tcx>(
                 ));
             }
 
+            let is_unsized_field = |field: &FieldDef| {
+                let field_ty = tcx.type_of(field.did);
+                tcx.try_instantiate_and_normalize_erasing_regions(args, cx.param_env, field_ty)
+                    .map(|f| !f.is_sized(tcx, cx.param_env))
+                    .map_err(|e| {
+                        error(
+                            cx,
+                            LayoutError::NormalizationFailure(field_ty.instantiate_identity(), e),
+                        )
+                    })
+            };
+
+            if def.is_struct()
```

Comment:
Will we now have the same issue with enums and unions?

Please add some tests for enums. I don't think unions are a problem without anonymous struct fields.

You can walk all variants and then iterate over their fields to handle enums, too

Reply:
I've added tests for enums and unions and they are passing. So looks like the ICE does not occur for them. Let me know if you want me to still add the logic to handle them in layout.

Reply:
It does fail for single-variant enums:

```rust
use std::mem::size_of;

enum BadEnum {
    Variant1 {
        field1: i32,
        field2: str, // Unsized
        //~^ ERROR the size for values of type `str` cannot be known at compilation time
        field3: [u8; 16],
    },
}

pub fn main() {
    let _a = &size_of::<BadEnum>();
    assert_eq!(size_of::<BadEnum>(), 21);
}
```

It may be possible to construct a multi-enum version, too, but I can't come up with it right now

Reply:
You're right. it does fail with single variant enums, but I too couldn't find a way to repro for multi-variant ones.

I've now added the logic to handle enums and relevant tests.